backend/src/main/java/com/google/sps/datastore/ParticipantDatastore.java

```diff
@@ -0,0 +1,118 @@
+    participantEntity.setProperty(PROPERTY_ENDTIMEAVAILABLE, participant.getEndTimeAvailable());
+    participantEntity.setProperty(PROPERTY_DURATION, participant.getDuration());
+    participantEntity.setProperty(PROPERTY_MATCHID, participant.getMatchId());
+    participantEntity.setProperty(PROPERTY_TIMESTAMP, participant.getTimestamp());
+
+    // Insert entity into datastore
+    datastore.put(participantEntity);
+  }
+
+  /** Return Participant Entity from username, or null if entity is not found */
+  @Nullable
+  private Entity getEntity(String username) {
+    Key participantKey = KeyFactory.createKey(KIND_PARTICIPANT, username);
+    try {
+      return datastore.get(participantKey);
+    } catch (EntityNotFoundException e) {
+      return null;
+    }
+  }
+
+  /** Return Participant from username, or null if participant with username not in datastore */
+  @Nullable
+  public Participant getParticipantFromUsername(String username) {
+    Entity entity = getEntity(username);
+    if (entity == null) {
+      return null;
+    }
+    return getParticipantFromEntity(entity);
+  }
+
+  /** Return participant object from datastore participant entity, or null if entity is null */
+  @Nullable
+  private static Participant getParticipantFromEntity(@Nonnull Entity entity) {
+    // Get entity properties
+    String username = (String) entity.getProperty(PROPERTY_USERNAME);
+    long startTimeAvailable = (long) entity.getProperty(PROPERTY_STARTTIMEAVAILABLE);
+    long endTimeAvailable = (long) entity.getProperty(PROPERTY_ENDTIMEAVAILABLE);
+    int duration = ((Long) entity.getProperty(PROPERTY_DURATION)).intValue();
+    long matchId = (long) entity.getProperty(PROPERTY_MATCHID);
+    long timestamp = (long) entity.getProperty(PROPERTY_TIMESTAMP);
+
+    // Create and return new Participant
+    return new Participant(
+        username, startTimeAvailable, endTimeAvailable, duration, matchId, timestamp);
+  }
+
+  /** Return list of all unmatched participants with same duration */
+  public List<Participant> getSameDurationParticipants(int duration) {
+    Query query = new Query(KIND_PARTICIPANT).addSort(PROPERTY_DURATION, SortDirection.ASCENDING);
+
+    // Create filter to get only participants with same duration
+    Filter sameDuration = new FilterPredicate(PROPERTY_DURATION, FilterOperator.EQUAL, duration);
+    query.setFilter(sameDuration);
+
+    PreparedQuery results = datastore.prepare(query);
+
+    // Convert entities to list of participants
+    List<Participant> participants = new ArrayList<Participant>();
+    for (Entity entity : results.asIterable()) {
+      participants.add(getParticipantFromEntity(entity));
+    }
+    return participants;
+  }
+
+  /** Remove Participant from datastore */
+  public void removeParticipant(String username) {
+    Key participantKey = KeyFactory.createKey(KIND_PARTICIPANT, username);
+    datastore.delete(participantKey);
```

Comment:
any exceptions we need to worry about here if the key isn't in the datastore when calling delete?

Reply:
If the key is invalid, an IllegalArgumentException is thrown.

Reply:
Up to you how you want to handle that. I could see an argument for throwing or just swallowing the exception here.

Reply:
What do you mean by swallowing the exception? This is another instance when removeParticipant will only be called if the participant actually is in the datastore, so I'm not sure what to return. Should I add a try catch, return a boolean, and send out an HTTP response error?

Reply:
My bad. To "swallow" an exception is to catch it so that it isn't thrown any further up the call stack. So this method could either throw the exception or just catch and log it. The difference is in how the application should behave when this happens. 

If the method throws the servlet will crash and if it catches the exception then it will continue as normal. You might be thinking crashing is always bad but it's also a clear indicator to us the developers that there's an error that we might be able to fix. 

For this specific case, I'm leaning towards catching and logging the exception without throwing anything to the caller. If two servlets want to delete the same entry from the datastore at the same time, there's no reason that one of them should crash. WDYT?

Reply:
I see, I changed it so it catches the exception and prints. I'm planning on investigating how to use the logger in a future PR.